DRODLib/Character.cpp

```diff
@@ -1288,6 +1321,43 @@ int CCharacter::parseFactor(const WCHAR *pwStr, UINT& index, CCurrentGame *pGame
 
 		//else: unrecognized identifier -- just return a zero value
 		return val;
+	} else if (pwStr[index] == W_t('#') || pwStr[index] == W_t('@')) {
+		//Find spot where var identifier ends.
+		int endIndex = index + 1;
+		int spcTrail = 0;
+		while (CDbHold::IsVarCharValid(pwStr[endIndex]))
+		{
+			if (pwStr[endIndex] == W_t(' '))
+				++spcTrail;
+			else
+				spcTrail = 0;
+			++endIndex;
+		}
+
+		const WSTRING wVarName(pwStr + index, endIndex - index - spcTrail);
+		const bool bIsLocalArray = ScriptVars::IsCharacterLocalVar(wVarName);
+		index = endIndex;
+
+		if (bIsLocalArray && !pNPC)
+			return 0;
+
+		ASSERT(pwStr[index] == W_t('['));
+		++index; //pass left bracket
+
+		int arrayIndex = parseExpression(pwStr, index, pGame, pNPC, ']'); //recursive call
+		SKIP_WHITESPACE(pwStr, index);
+		if (pwStr[index] == W_t(']'))
+			++index;
+		else
+		{
+			//parse error -- return the current value
+			LogParseError(pwStr, "Parse error (missing close bracket)");
+		}
+
+		UINT varId = pGame->pHold->GetVarID(wVarName.c_str());
+		ScriptArrayMap& scriptArrays = bIsLocalArray ? pNPC->localScriptArrays : pGame->scriptArrays;
+		int val = scriptArrays[varId][arrayIndex];
```

Comment:
Here, I think we need to check whether varId is an array var or not.
If not, val should probably default to 0. Otherwise, we're creating a new empty map here, which is wasteful. Only look at pre-existing maps.

I think we should also check whether [arrayIndex] is present in an existing map via find(). If not, return 0. Otherwise, dereference the iterator and return the value. (Don't use operator[arrayIndex] after a find(), or this will wastefully traverse the map structure a second time to locate the key value again!)